src/platform/common/utils/async.ts

```diff
@@ -244,3 +250,1793 @@ export class PromiseChain {
         return deferred.promise;
     }
 }
+
+export function isThenable<T>(obj: unknown): obj is Promise<T> {
+    return !!obj && typeof (obj as unknown as Promise<T>).then === 'function';
+}
+
+export interface CancelablePromise<T> extends Promise<T> {
+    cancel(): void;
+}
+function createCancellationError() {
+    class InternalCancellationError extends Error {
+        constructor() {
+            super('Cancelled');
+            this.name = 'InternalCancellationError';
+        }
+    }
+    try {
+        if (require('vscode').CancellationError) {
+            return new (require('vscode').CancellationError)();
+        } else {
+            return new InternalCancellationError();
+        }
+    } catch {
+        return new InternalCancellationError();
+    }
+}
+export function createCancelablePromise<T>(callback: (token: CancellationToken) => Promise<T>): CancelablePromise<T> {
+    const source = new CancellationTokenSource();
+
+    const thenable = callback(source.token);
+    const promise = new Promise<T>((resolve, reject) => {
+        const subscription = source.token.onCancellationRequested(() => {
+            subscription.dispose();
+            source.dispose();
+            reject(createCancellationError);
+        });
+        Promise.resolve(thenable).then(
+            (value) => {
+                subscription.dispose();
+                source.dispose();
+                resolve(value);
+            },
+            (err) => {
+                subscription.dispose();
+                source.dispose();
+                reject(err);
+            }
+        );
+    });
+
+    return <CancelablePromise<T>>new (class {
+        cancel() {
+            source.cancel();
+        }
+        then<
+            TResult1 = T,
+            TResult2 = never
+        >(resolve?: ((value: T) => TResult1 | Promise<TResult1>) | undefined | null, reject?: ((reason: any) => TResult2 | Promise<TResult2>) | undefined | null): Promise<TResult1 | TResult2> {
+            return promise.then(resolve, reject);
+        }
+        catch<
+            TResult = never
+        >(reject?: ((reason: any) => TResult | Promise<TResult>) | undefined | null): Promise<T | TResult> {
+            return this.then(undefined, reject);
+        }
+        finally(onfinally?: (() => void) | undefined | null): Promise<T> {
+            return promise.finally(onfinally);
+        }
+    })();
+}
+
+/**
+ * Returns a promise that resolves with `undefined` as soon as the passed token is cancelled.
+ * @see {@link raceCancellationError}
+ */
+export function raceCancellation<T>(promise: Promise<T>, token: CancellationToken): Promise<T | undefined>;
+
+/**
+ * Returns a promise that resolves with `defaultValue` as soon as the passed token is cancelled.
+ * @see {@link raceCancellationError}
+ */
+export function raceCancellation<T>(promise: Promise<T>, token: CancellationToken, defaultValue: T): Promise<T>;
+
+export function raceCancellation<T>(
```

Comment:
These are the things I want in this repo, a lot of these helper classes/functions have been re-invented in Jupyter, 
Its simpler to use the same code that is in VS Code (easier for everyone and its also better)